### PR TITLE
feat(lcma): LLM-synthesis substrate — provider client, budget + inference ledgers (WS1 PR1)

### PR DESCRIPTION
**Pull request by davesnowdon**
## What & why

First of two merge-gated PRs delivering **MVP-3 WS1 slice 1** (LLM-synthesis backbone + thin typed-edge inference; tracker task `7387506b`, epic `d921d168`, plan approved 2026-07-25). This PR is the substrate — **no behavior change**: nothing constructs the client yet. PR2 wires it into the enrich worker as the typed-edge inference consumer.

Design context: the 2026-07 substrate audit found the typed-edge graph semantically empty (17% coverage, 2 genuine relation types) — LLM typed-edge inference is the top MVP-3 lever, and this backbone is its prerequisite (`docs/plans/lcma-design.md` WS1/WS2).

## Changes

1. **Config** — `LcmaConfig.llm`: nested `LlmConfig` for one generic OpenAI-compatible chat-completions endpoint (covers Ollama/llama.cpp/vLLM `/v1` shims locally — privacy-first stays whole — and hosted APIs). Enabled ⇔ `base_url` set (kill switch = unset). `SecretStr` api_key hardened end-to-end: outer whitespace stripped (file-backed secrets), header-unsafe values (control chars incl. CR/LF, non-ASCII, interior spaces) and non-string types rejected with a **value-free `ConfigurationError`** — never pydantic-wrapped, so no `.errors()`/`.json()` surface can echo the value; the client additionally sanitizes transport-error messages (h11 `LocalProtocolError` reduced to its type name, key substring redacted elsewhere). Other validators: model required when enabled; similarity band ordered. **Replaces the dead `llm_provider` field** (never read by any code) — lean removal, no shim. Env: `LITHOS_LCMA__LLM__BASE_URL` etc. (nested delimiter already supported).
2. **Client** — `lithos/lcma/llm.py` (~150 LOC, no SDK): async httpx POST to `{base_url}/chat/completions`; Bearer auth when key set; `response_format=json_object` (maps to Ollama `format=json`; callers still parse defensively); one retry on timeout/connect/5xx, immediate fail on 4xx; every failure → `LlmError` (new, under `CognitiveMemoryError`). Token accounting from the `usage` block with a flagged chars//4 estimate when the endpoint omits it.
3. **stats.db ledgers** (additive `CREATE TABLE IF NOT EXISTS`, no migration): `llm_budget` — UTC-day token/call ledger with atomic `ON CONFLICT` upsert (race-free spend recording); `edge_inference_log` — idempotency ledger keyed by the identity triple **`(node_id, doc_version, inference_version)`**: a re-drained unchanged node skips in O(1); bumping the engine-owned `INFERENCE_VERSION` (prompt/vocab/parse changes; wired in PR2) reprocesses the corpus; a model switch deliberately does **not** invalidate — the operator reset is `purge_edge_inference_log(model=...)`, which clears execution markers only (previously inferred edges remain; reconciliation of superseded inferred edges is slice-2 sweep scope).
4. **Telemetry** — `lcma_llm_calls{outcome}`, `lcma_llm_tokens{estimated}`, `lcma_llm_call_duration` (ms), `lcma_edge_inference_skips{reason}`, `lcma_inferred_edges_written{relation}` (lazy properties, NoOp-safe).
5. **Deps** — `httpx>=0.27` promoted to main dependency: already transitively installed via fastmcp + chromadb, so zero image cost; graceful degradation is a config concern (`base_url=None`), not an install-state concern. `uv.lock` regenerated.

## Behavior note (review round 3/4)

Invalid legacy `LITHOS_PORT` now raises `lithos.errors.ConfigurationError` instead of a `ValueError`. This is intentional and load-bearing: pydantic's `ValidationError` *is* a `ValueError` subclass, so keeping the old catchable type and closing the structured-error secret leak are mutually exclusive — any pydantic-wrapped error resurrects the `.errors()`/`.json()` surface that echoes raw config input. Library callers catching `ValueError` around `load_config()` should catch `ConfigurationError` (or `LithosError`). CLI UX is unchanged (an uncaught traceback either way); a concise Click-level rendering is deliberate follow-up work.

## Test plan

- `make check` green — **1732 passed** (64 new tests after four review rounds: header-unsafe/malformed api_key rejection with value-absence assertions, sanitized transport-error surfaces, enclosing-validator secret redaction (field-level validation on LcmaConfig + non-pydantic `ConfigurationError` for legacy `LITHOS_PORT`), secret redaction incl. structured errors()/json(), LlmError boundary sweep, usage-value edge cases, concurrent ledger, pre-WS1-db additive migration, instrument metadata on the real SDK path).
- New: `tests/test_llm_client.py` (httpx.MockTransport: request shape incl. Bearer + response_format; usage extraction; estimation fallback; 5xx retry-then-succeed; persistent timeout → LlmError; 4xx no-retry; malformed body). `test_stats.py`: budget accumulation/day-isolation/clamping; op-log version semantics + idempotency; schema table lists updated. `test_lcma_config.py`: enabled semantics, validators, SecretStr redaction, `LITHOS_LCMA__LLM__*` env round-trip.
- `docs/generated/` regenerated (LOC drift) and committed; SPECIFICATION.md stats.db table list updated.
- Integration suite unaffected (no wiring yet); will run on PR2.

## Not in this PR (PR2, stacked next)

`edge_inference.py` consumer, enrich-worker wiring (+`search` dep), `intake.assert_edge` origin stamp (cascade guard), integration test, checklist/design-doc ticks.

### CODE REVIEW

**Comment by davesnowdon** [comment]
Thanks — both blockers were real and are fixed, plus the recommended items. All in 466eef5.

**1. API-key leak in validation errors (High) — fixed, and your propagation warning was the actual bug.** `hide_input_in_errors=True` on `LlmConfig` alone was NOT enough: Pydantic formats a `ValidationError` with the config of the **outermost** validating model, so a failure reached through `LithosConfig` still echoed the unwrapped key (reproduced exactly as you predicted before fixing). The flag is now set on every construction entry point — `LlmConfig`, `LcmaConfig`, and the `LithosConfig` settings root — with a comment explaining why it must live at the root. Two new tests: direct construction with a failing sibling field, and the env-driven path through `LithosConfig()`; both assert the key is absent from `str()`/`repr()` of the error. The trade-off (config validation errors no longer echo raw input tree-wide) is deliberate: the field name + constraint remain, and this tree carries secrets.

**2. LlmError boundary (Medium) — fixed.** The transport catch is now `httpx.HTTPError` (the base of every transport/protocol failure — ReadError, WriteError, RemoteProtocolError, UnsupportedProtocol, invalid URLs included), so nothing escapes the contract. All of them get the single retry — one wasted 0.5s pause on a non-transient failure is cheaper than misclassifying a transient one; 4xx still fails immediately. Usage parsing: a new `_usage_int` helper requires **both** fields to be valid non-negative ints (bool explicitly excluded) or the client falls back to the flagged chars//4 estimate — no raw `TypeError` escapes and bogus usage can't corrupt the ledger. New tests: ReadError, RemoteProtocolError, and a parametrized sweep over null / non-numeric / negative / missing-completion usage.

**3. Docs (Medium) — fixed in this PR.** The reference-updates were sitting in the stacked #406; you're right they belong in the PR that removes the field. This PR now updates every active `llm_provider` reference (checklist lines 56 + WS1 line, design WS1 paragraph, LLM-integration-scope section, §7.z draft schema) to the nested `lcma.llm` schema incl. env-var form. #406 will rebase and keep only its own DONE-tick annotations. ADR-0008's mention stays as-is — it records a historical decision about a rejected alternative.

**4. Budget semantics (Medium) — documented as soft.** Answering your question directly: `daily_token_budget` is a **soft stop checked before starting the next call**, with bounded overshoot of at most the final call's tokens. That is now stated on the config field and on `record_llm_spend`'s docstring. Multi-worker: not supported — the architecture is one in-process `EnrichWorker` per server process, one server per stats.db; spend *recording* stays atomic regardless, only admission is soft. A hard reservation model isn't warranted for a local-Ollama cost profile; if a hosted-API deployment ever needs it, that's a follow-up with a pessimistic-reserve/reconcile ledger.

**5. Empty/whitespace endpoint (Low) — fixed.** `base_url` normalises `""`/whitespace → `None` (disabled) in a `before` validator — an empty env var now means "off", not "fail every call". Enabled URLs must be `http://`/`https://`; `model` must be non-blank when enabled. Tests for all three.

**Also picked up from your list:** inference-ledger test now verifies the persisted `model`/`edges_written`/`inferred_at` audit fields and first-write-wins semantics; NoOp construction+record test for all five new instruments; `json_response=False` request-shape test.

**Copilot's two comments are also addressed:** the empty/whitespace `base_url` one is the same as finding 5; the `usage`-shape one is fixed — a `usage` block carrying only `total_tokens` (or a broken split alongside a valid total) now uses the total as **measured** spend (it's exactly what the budget ledger consumes) with the split recorded as total/0, instead of being mis-flagged as estimated. Two new tests.

`make check` green on the head (1705 passed).

**Comment by davesnowdon** [comment]
All three blockers fixed in 7848662, plus the recommended coverage. Answers inline.

**1. Structured error leak (High) — fixed at the root this time.** You were right that `hide_input_in_errors` only redacts printed forms; I reproduced the `.errors()[0]["input"]["api_key"]` / `.json()` leak on both construction paths before fixing. The fix removes the *cause* rather than filtering the symptom: **all `LlmConfig` validation is now field-level** (base_url normalize+scheme, model-required via `ValidationInfo.data` with `validate_default=True`, similarity-band on `max_similarity`) — there is no model validator left, so no error entry ever carries the whole input dict; a field error's `input` is only that field's own value. `hide_input_in_errors` stays as printed-form defense in depth. Tests now assert `sk-topsecret` absent from `str()`, `repr()`, `str(errors())`, **and** `.json()` on all three paths: direct `LlmConfig`, nested `LcmaConfig(llm={...})`, and env-driven `LithosConfig()`. (Q1: yes — structured coverage is now the contract, and tested.)

**2. `_usage_int` (Medium) — fixed.** Accepts ints and *integral finite* floats (JSON decoders may deliver `120.0` — accepted explicitly, per your either/or); fractional, NaN, and ±infinity return `None` → the flagged estimate path. Nothing can raise past the boundary. Parametrized tests for `1.9`, NaN, ±inf on the split fields and on `total_tokens` (response bodies built as raw text since httpx's strict `json=` refuses to encode non-finite tokens that real decoders accept), plus a test that `120.0/30.0` is measured-not-estimated. (Q2: fractional → estimated; integral floats accepted.)

**3. Ledger invalidation (Medium) — decided and encoded before the schema ships.** The identity is now the triple **`(node_id, doc_version, inference_version)`**:
- `inference_version` is the inference engine's code-level policy fingerprint (prompt schema, relation vocab, parse semantics — a constant in `edge_inference.py`, wired in PR2). Bumping it on a prompt/policy change naturally reprocesses unchanged notes.
- A **model/config switch deliberately does not invalidate** — flipping `LITHOS_LCMA__LLM__MODEL` must not silently re-burn the whole corpus budget. The supported reset for that is the new **`StatsStore.purge_edge_inference_log(model=...)`** (per-model or full), after which re-inference upserts over existing edges. The slice-2 sweep backfill will consult `inferred_at` for age-based re-inference on top of this.
(Q3: permanent per identity-triple; purge is the supported re-inference workflow; both semantics are now docstring'd on `has_edge_inference`.)

**4. Concurrent ledger test (Low) — added.** Two `StatsStore` instances (separate connections) gather 50 interleaved `record_llm_spend` upserts against one db; exact totals asserted from both stores — the single-statement `ON CONFLICT` upsert is the atomicity boundary, not the per-store lock. Also added the pre-WS1-database test: drop both new tables from a populated db, reopen, assert the ledgers are recreated and existing `node_stats` rows survive. (Q4: single sequential worker per stats.db is an architectural constraint — the worker is in-process in the one server process that owns the db, one container per environment — not a cross-process lock; spend *recording* is atomic regardless, as the new test shows.)

**5. Telemetry metadata (Low) — added.** `TestLlmInstrumentMetadata` drives the five instruments through a real SDK `MeterProvider` + `InMemoryMetricReader` and asserts names, counter-vs-histogram kinds (monotonic sums checked), the `ms` unit on the duration histogram, values, and attribute sets.

**Also picked up:** non-empty `base_url` values are whitespace-stripped; stats.py's module docstring now lists all tables incl. the two WS1 ledgers; PR description test counts refreshed.

`make check` green on the head (1718 passed). PR2 (#406) will rebase on this and wire `INFERENCE_VERSION` through the engine.

**Comment by davesnowdon** [comment]
Round-3 blocker fixed at the ancestor level; both repros now covered by regression tests. Answers inline.

**1. Enclosing-validator leak (High) — fixed.** You identified the last two model validators above the secret subtree; both are gone as leak surfaces:
- `LcmaConfig`'s two dict-shaping validators (`rerank_weights`, `note_type_priors`) are now **field validators** (`mode="after"`) — a failure's error `input` is scoped to that field's own dict, never the sibling `llm` subtree. Behavior (fill-missing, renormalize, unknown-key rejection) is unchanged and the pre-existing tests confirm it; the conversion also stops mutating the caller's dict in place.
- `LithosConfig._apply_backward_compat_env_overrides` stays a model validator (it needs `model_fields_set` for the constructor-wins semantics) but can no longer raise *through pydantic*: bad `LITHOS_PORT` now raises a new **`ConfigurationError(LithosError)`** — deliberately not a `ValueError`, so it propagates unwrapped with only its message; there is no `ValidationError`, hence no `.errors()`/`.json()` surface at all. The invariant is documented on the validator, on `LcmaConfig`, and on the exception class: *model validators at/above the secret subtree must never raise ValueError*.
- Regression tests (all three you listed): valid key + unknown `rerank_weights` key; valid key + unknown `note_type_priors` key; env key + `LITHOS_PORT=not-a-port` (asserts `ConfigurationError`, not `ValidationError`, and key absent from `str`/`repr`). The sibling tests reuse `_assert_key_fully_absent` across `str`/`repr`/`errors()`/`json()`.

(Q1: yes — the contract is that **no configuration failure anywhere in the tree** exposes the key through any `ValidationError` surface, structured or printed. After this change the only validators that can observe the raw `lcma` subtree cannot produce a `ValidationError` at all.)

**2. Re-inference reconciliation (Medium) — decided: additive in this slice, reconciliation is slice-2 scope.** `purge_edge_inference_log` clears *execution markers only*; its docstring now says exactly that, including your scenarios (a relation/direction change keys a new edge alongside the old one rather than replacing it). Rationale for not reconciling here: PR1 has no consumer and no edge-diff machinery, and the slice-2 sweep/backfill is the natural home — it will diff existing `provenance_type="inferred"` edges against a fresh adjudication before deciding what to retire, rather than blind-deleting on purge (which would also destroy edges a *correct* older model asserted). Kept the name (it does purge the log — the docstring now prevents the stronger reading), per your nice-to-have framing. (Q2: deliberately accumulate in slice 1; supersession handled by the sweep policy in slice 2, tracked in the design doc's WS2 section.)

**3. PR description (Low) — updated.** The ledger paragraph now describes the `(node_id, doc_version, inference_version)` triple, the `INFERENCE_VERSION` bump semantics, the deliberate model-switch non-invalidation, and purge-as-marker-reset.

(Q3: slice 2 ships the operator surface — the sweep/backfill CLI (`python -m lithos.cli`, alongside `recalibrate-salience`) is where purge + re-inference belong operationally; until then it is intentionally not reachable from any tool surface, matching "PR1 has no behavior change".)

Also folded in: the code-quality bot's mixed-import-style fix in `tests/test_telemetry.py`.

`make check` green on the head (1721 passed, +3 regression tests).

File: tests/test_telemetry.py [hunk 1]
```diff
@@ -1582,3 +1602,63 @@ def test_reset_for_testing_resets_lcma_flag(self):
...
+)
+class TestLlmInstrumentMetadata:
+    """Round-2 review: verify the WS1 instruments' names, kinds, units, and
+    attributes on the real SDK path, not just NoOp construction."""
+
+    def test_llm_instruments_emit_expected_metadata(self, monkeypatch) -> None:
+        from opentelemetry.sdk.metrics import MeterProvider
+        from opentelemetry.sdk.metrics.export import InMemoryMetricReader
+
+        import lithos.telemetry as telemetry_mod
```
[the remark this answers is not in the document]
Comment (davesnowdon) [reply tests/test_telemetry.py]: Adopted: the test now uses `from lithos.telemetry import lithos_metrics` and patches by string target (`monkeypatch.setattr("lithos.telemetry.get_meter", ...)`), so this PR's diff no longer adds any module-style import of `lithos.telemetry`.

Note for the record: this rule is file-scoped and `tests/test_telemetry.py` has mixed both import styles since before this PR (~15 pre-existing `import lithos.telemetry as tel` sites alongside dozens of `from lithos.telemetry import ...`), so the module-level alert may persist against baseline code. Harmonizing the whole file is out of scope for this PR.

File: src/lithos/lcma/llm.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    call :meth:`close` on worker shutdown.
+    """
+
+    def __init__(self, config: LlmConfig, *, transport: httpx.AsyncBaseTransport | None = None):
+        if config.base_url is None:
+            raise ValueError("LlmClient requires lcma.llm.base_url to be set")
+        self._config = config
+        headers = {}
+        if config.api_key is not None:
+            headers["Authorization"] = f"Bearer {config.api_key.get_secret_value()}"
```
[the remark this answers is not in the document]
Comment (davesnowdon) [reply src/lithos/lcma/llm.py]: False positive — the header is built from the SecretStr's unwrapped value, not a literal:

```python
headers["Authorization"] = f"Bearer {config.api_key.get_secret_value()}"
```

The `******` you're seeing is `SecretStr`'s masked `repr`, which is exactly the point of using it (the key masks everywhere *except* the explicit `get_secret_value()` call at the transport boundary). `test_bearer_header_sent_when_api_key_set` proves the real header reaches the wire: it drives the client through `httpx.MockTransport` and asserts the received request carries `Authorization: Bearer sk-secret`.

**Comment by davesnowdon** [comment]
Both blockers fixed — at the config boundary first, with the client hardened as defense in depth. Answers inline.

**1. Header-unsafe keys leaking via h11/httpx (High) — fixed at two layers.**
- *Config (root fix):* new `mode="before"` field validator on `api_key` makes header-unsafe keys **unrepresentable in `LlmConfig`**: embedded CR/LF, control characters, non-ASCII, and interior spaces are rejected; only visible ASCII passes. Your realistic case — a file-backed secret with a trailing newline — now *works* instead of leaking: outer whitespace is stripped (matching `base_url`), and empty/whitespace-only normalises to `None`. Since the client can only be constructed from a validated `LlmConfig`, the `LocalProtocolError` and `UnicodeEncodeError` paths can no longer be reached with a credential in them.
- *Client (your suggested defense in depth):* `_sanitize_error()` now mediates every message that reaches logs or `LlmError`: `httpx.LocalProtocolError` is reduced to its type name (h11 embeds the raw `Authorization` bytes in the message), and **its cause chain is severed** (`from None`) so no traceback formatter can print them either; every other transport error has the key substring redacted (`[redacted]`) in case a proxy echoes request data.
- *Regression tests:* both layers — config rejection is parametrized over CR, LF, NUL, non-ASCII, and interior space (asserting the exception is *not* a `ValidationError` and the value is absent from `str`/`repr`); client tests drive a synthetic `LocalProtocolError` carrying `Bearer sk-secret` and a key-echoing `ConnectError` through MockTransport, asserting the secret is absent from the `LlmError` message, `__cause__ is None`, and — via caplog — the retry warning.

(Q1: yes, now intentionally unsupported *and enforced* — visible ASCII only, outer whitespace normalised away. Documented on the validator.)

**2. Malformed-type api_key via `.errors()`/`.json()` (Medium) — fixed by the same validator.** Non-`str`/`SecretStr`/`None` types (your YAML-list shape) raise a value-free `ConfigurationError` *before* pydantic can capture the value as the field-error `input` — there is no `ValidationError`, hence no structured surface at all. Same for the charset rejections in (1): they must not be `ValueError` either, because a field-validator `ValueError` puts the offending *value itself* into `errors()[0]["input"]`. Tests cover the direct and nested (`LcmaConfig(llm={"api_key": [...]})`) paths.

**3. ValueError → ConfigurationError (Low) — intentional, now documented in the PR description.** The two goals are mutually exclusive: pydantic's `ValidationError` *is* a `ValueError` subclass, so any error that stays catchable as `ValueError` via pydantic wrapping resurrects the `.errors()`/`.json()` surface. The PR description now carries a behavior note telling library consumers to catch `ConfigurationError`/`LithosError`. CLI UX is unchanged in practice — this misconfiguration previously surfaced as an uncaught `ValidationError` traceback, now an uncaught `ConfigurationError` traceback; a concise Click-level rendering is a sensible follow-up and noted as such. (Q2: yes, intended.)

(Q3: as per round 3 — the slice-2 sweep/backfill CLI (alongside `recalibrate-salience`) is the operator surface for purge + re-inference; until then `purge_edge_inference_log` is internal/direct use only, and PR2's integration tests will keep the ledger-reset vs edge-reconciliation distinction explicit.)

`make check` green on the head. On the MockTransport limitation you noted: agreed it can't exercise h11's real serialization — that's exactly why the fix rejects unsafe keys at config time (provable without a socket) and treats client-side sanitization as belt-and-braces rather than the boundary.

**Comment by davesnowdon** [comment]
Heads-up: that pass reviewed a stale checkout — `51e3c7d` predates the round-4 fix. The api_key hardening is in **`ee14e43`** (current branch head, pushed with the round-4 response comment above).

Verified both of your repros on `ee14e43`:

- `LlmConfig(api_key=["sk-secret"])` → raises value-free `ConfigurationError` (not a `ValidationError` — no `.errors()`/`.json()` surface exists); the value appears in neither `str` nor `repr`.
- A key with an embedded newline → same value-free `ConfigurationError` at config time, so it can never reach `LlmClient`, h11, or the retry warning. A *trailing* newline (the file-backed-secret case) is stripped and the key works. As defense in depth the client also reduces `httpx.LocalProtocolError` to its type name (message and severed cause chain) and redacts the key substring from all other transport-error text — see `test_local_protocol_error_never_echoes_header_bytes` / `test_transport_error_message_redacts_api_key` and the parametrized `test_header_unsafe_api_key_rejected_value_free`.

Full detail in the round-4 response above; `make check` green at `ee14e43` (1732 passed).

File: src/lithos/lcma/llm.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+                response = await self._client.post("/chat/completions", json=body)
+            except httpx.HTTPError as exc:
+                last_error = exc
+            else:
+                if response.status_code < 500:
+                    if response.status_code >= 400:
+                        raise LlmError(
+                            f"LLM endpoint returned {response.status_code}: {response.text[:200]}"
+                        )
+                    return response
```
[the remark this answers is not in the document]
Comment (davesnowdon) [reply src/lithos/lcma/llm.py:163]: Good catch — this was the one error surface the round-4 sanitization pass missed. Fixed: the 4xx path now runs the response body through the same redaction as `_sanitize_error()` (extracted as a shared `_redact()` helper), and it redacts **before** the 200-char truncation so a key straddling the cut can't survive as an unmatchable partial. Regression test: a 401 body echoing `Bearer sk-secret` produces an LlmError containing `[redacted]` and no key (`test_4xx_body_echoing_api_key_is_redacted`).

**Comment by davesnowdon** [comment]
Both closed — you and Copilot converged on the 4xx body independently, and the `__cause__` point is exactly right: my round-4 fix severed the chain only for `LocalProtocolError`, but `traceback.format_exception()` renders *any* cause's raw args.

**1. Unsanitized `__cause__` — fixed.** `_post_with_retry` now raises `LlmError ... from None` unconditionally: no transport exception ever rides the cause chain, so there is nothing for a traceback formatter to render. Diagnostics are preserved in the sanitized message itself (original exception type name + redacted detail). Your observation about the ConnectError test was correct — it asserted the message but never inspected the chain; all three leak tests now assert `__cause__ is None` **and** run `"".join(traceback.format_exception(exc))` asserting the key is absent from the full rendered traceback, per your suggestion.

**2. 4xx body — fixed.** The 4xx path routes the response body through the shared `_redact()` helper before embedding it, and redacts **before** the 200-char truncation so a key straddling the cut can't survive as an unmatchable partial. Regression test: a 401 body echoing `Bearer sk-secret` yields an LlmError containing `[redacted]`, with the same format_exception assertion.

(The `_parse_response` chain (`from exc` on JSON/lookup errors) is deliberately kept: those causes are decode-position messages and key names from the *response* parse — no request-side data — and the chaining is useful when a model emits garbage.)

`make check` green on the new head.
